### PR TITLE
WIP Adds entity type registration service

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds a more versatile entity type registration service that can be
used to register entity types for specific contexts, as well as
to dynamically define entity subtype class

### CODE REVIEW

**Comment by hypeJunction** [comment]
In case of #9918, plugins could just do:

```php
elgg_register_entity_type('object', 'blog', ElggBlog::class);

// later on
elgg_register_entity_type('object', 'blog', null, ['river' => true]);
```

My thinking we could eventually use this approach to configure form fields as well.

File: engine/tests/phpunit/Elgg/lib/entities/EntityTypeRegistrationTest.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+namespace Elgg\lib\entities;
+
+/**
+ * @group Entities
+ */
+class EntityTypeRegistrationTest extends \Elgg\TestCase {
+
+	public function setUp() {
+		$mocks = new \Elgg\Tests\EntityMocks($this);
+		unset(_elgg_services()->entityTypes); // previous state is floating over
```
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/lib/entities/EntityTypeRegistrationTest.php]: @mrclay This stuff is bugging me. Maybe we should go back to your original proposal of adding a method to clear/backup `_elgg_services()`. Test cases should run independently from each other, and this stuff keeps popping up and hard to figure out why.

**Comment by mrclay** [comment]
A lot to consider here. I'd like to separate PHP class association from "registrations" for search and other Elgg features; these are very different purposes and I think a tiny bit more boilerplate is worth it.

add_subtype() is already simple to understand and I have a PR that just makes that something you'd move to my_init(), no activate.php required.

The context part looks interesting, though they're pretty mysterious. Sometimes they mean the first URL segment, sometimes not. I'd rather register a subtype with a list of named "features" instead. I guess just don't call it context is my suggestion.

**Comment by mrclay** [comment]
I don't like the volatileData injection. We're really talking about a static thing, not instance specific. Adding a static function to ElggEntity would make more sense and avoid issues with caching in memcache/serialization.

**Comment by hypeJunction** [comment]
I am fine with just using `features` as an flat array and avoiding volatile data injection.

**Comment by hypeJunction** [comment]
I will wait for #8688 to wrap this one up. Just putting it out there as something to consider - I have been hatching this egg for a while now.

**Comment by iionly** [comment]
Being able to define contexts for subtypes would be part of the way to solve the problems I have described in https://github.com/Elgg/Elgg/issues/9918. I think for the search context it's fine to just say yes/no regarding a subtype to be considered.

But the other issue is with the river filter where a simple yes/no might not be enough in some cases. One example I'm struggling with is Tidypics (maybe not a good general example) where 3 different subtypes are used while I would like to not have 3 separate filter entries for these at all but would prefer one. While elgg_get_river() would allow combining them using subtypes the river.php view gives no option to alter the filter entries apart from just registering / not registering them both with the present implementation and this new implementation. Maybe the only way to be able to better configure the filter menu would be a handling the entries like a "river filter" menu and adding a plugin hook in the river view to allow for configuring the options to be used for elgg_list_river() for each registered menu item.

**Comment by mrclay** [comment]
There's a general need for a way to create custom filters for rivers *and* entity lists. Seems like an area where Drupal Views has this figured out, at least in flexibility.

**Comment by hypeJunction** [comment]
River filter could become a regular menu

**Comment by hypeJunction** [comment]
For lists, I have this https://github.com/hypejunction/elgg-object_sort

**Comment by hypeJunction** [comment]
Not going to see this through

**Comment by hypeJunction** [comment]
One thing to keep is a situation we have with discussion topics and replies. Replies are not registered as public facing entities, even though they are. They are searched for in bulk with topics.

**Pull request closed without merging**